### ISSUE TITLE
Support finding elements by View tag on Android

### ISSUE THREAD

**Issue by bootstraponline** [appium/appium#6025]
Tags enable dynamically storing information on an element that isn't read via screen readers. React native uses these to [store the test identifiers.](https://github.com/facebook/react-native/blob/90d2787aefbed27858f639e728882cdf784a0ec2/ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java#L72~L80) See this [appium discuss thread](https://discuss.appium.io/t/react-native-ui-element-access-via-testid/7845/4)

http://developer.android.com/reference/android/view/View.html#setTag(java.lang.Object)
http://developer.android.com/reference/android/view/View.html#getTag(int)
http://developer.android.com/reference/android/view/View.html#findViewWithTag(java.lang.Object)

>  It is generally preferable to use IDs (through the android:id attribute) instead of tags because they are faster and allow for compile-time type checking.

**Comment by talarari** [comment]
is this currently possible? or only in a future release?

**Comment by bootstraponline** [comment]
> is this currently possible?

nope

**Comment by jsdevel** [comment]
FYI, I have a PR open against react-native that will support `findViewById` with `testID`.  Check it out and up vote: https://github.com/facebook/react-native/pull/9942

**Comment by jlipps** [comment]
Some helpful code we might be able to use as a model: http://pastebin.com/h5TRLmL9

(This apparently comes from functionality FB added to Selendroid):


```java
package com.mypackage;

import android.util.Log;
import android.view.View;

import com.android.internal.util.Predicate;
import io.selendroid.server.common.Response;
import io.selendroid.server.common.SelendroidResponse;
import io.selendroid.server.common.http.HttpRequest;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Finds all views by given {@code View.getTag()}.
 */
public class FindElementsByAndroidTagHandler extends FindElementsBaseHandler {

  public FindElementsByAndroidTagHandler(String mappedUri) {
    super(mappedUri);
  }

  @Override
  public Response safeHandle(HttpRequest request) throws JSONException {
    JSONObject payload = getPayload(request);
    String tag = payload == null ? null : payload.getString("android_view_tag");

    String rootElementId = null;
    if (payload != null && payload.has("root_element_id")) {
      rootElementId = payload.getString("root_element_id");
    }

    Log.w(SelendroidExtension.TAG, "Getting views by tag: " + tag);

    if (tag == null || tag.isEmpty()) {
      return new SelendroidResponse(getSessionId(request), new JSONArray());
    }

    if (rootElementId != null) {
      return this.handleScopedFindElements(request, rootElementId, tag);
    }
    return this.handleGlobalFindElements(request, tag);
  }

  private Response handleGlobalFindElements(
      HttpRequest request,
      String tag) throws JSONException {
    return findElements(
        request,
        new ViewTagPredicate(tag));
  }

  private Response handleScopedFindElements(
      HttpRequest request,
      String elementId,
      String tag) throws JSONException {

    Log.w(SelendroidExtension.TAG, "Getting views as descendants of element with id: " + elementId);

    return findElements(
        request,
        elementId,
        new ViewTagPredicate(tag));
  }

  private static class ViewTagPredicate implements Predicate<View> {
    private String mTag;
    public ViewTagPredicate(String tag) {
      mTag = tag;
    }

    @Override
    public boolean apply(View view) {
      Object viewTag = view.getTag();
      return viewTag != null && viewTag.equals(mTag);
    }
  }
}
```

**Comment by jlipps** [comment]
@sravanmedarapu have you had a chance to look into this at all? would love to take a swing at it this sprint and @dpgraham is happy to help if need be.

**Comment by jasonhartley** [comment]
I eagerly await.  :-)

**Comment by dpgraham** [comment]
Question for @sravanmedarapu (or anyone who works on appium-uiautomator2-server)

Have you ever needed to get a reference to the main Activity using uiautomator? 

I tried doing this:

```java
Activity mainActivity = (Activity) InstrumentationRegistry.getContext();
```

But this doesn't allow the context to be cast to an activity.

I've also tried copying over Selendroid code (see https://github.com/selendroid/selendroid/blob/master/selendroid-server/src/main/java/io/selendroid/server/android/ViewHierarchyAnalyzer.java) which gets the top level views. But when I use it in my test app, it returns no views.

I need to be able to get access to the Activity (or the root View) so I can called 'findViewByWithTag' on it.

**Comment by bootstraponline** [comment]
> Activity mainActivity = (Activity) InstrumentationRegistry.getContext();

I think you want:

> InstrumentationRegistry.getTargetContext()

**Comment by dpgraham** [comment]
I tried that as well and it caused a casting error

```
java.lang.ClassCastException: android.app.ContextImpl cannot be cast to android.app.Activity
```

**Comment by bootstraponline** [comment]
> I tried that as well and it caused a casting error

You need something like [getCurrentActivity](https://gist.github.com/tir38/1a1c479b1f61e8b5b76c63eaadbf5dd4). Espresso makes this easy. Selendroid has their own implementation.

**Comment by dpgraham** [comment]
Thanks. I'm using this code snippet right now.

**Comment by dpgraham** [comment]
I've tried several different ways of getting the main activity with little luck.

When I attempt this:

```java
@Nullable
    public static Activity getCurrentActivity() {
        final Activity[] currentActivity = {null};
        getInstrumentation().runOnMainSync(new Runnable() {
            @Override
            public void run() {
                Collection<Activity> preOnCreateActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(PRE_ON_CREATE);
                Collection<Activity> createdActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(CREATED);
                Collection<Activity> startedActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(STARTED);
                Collection<Activity> resumedActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(RESUMED);
                Collection<Activity> pausedActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(PAUSED);
                Collection<Activity> stoppedActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(STOPPED);
                Collection<Activity> restartedActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(RESTARTED);
                Collection<Activity> destroyedActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(DESTROYED);
                Logger.info("!!!pre on create activities count ", preOnCreateActivities.size());
                Logger.info("!!!created activities count ", createdActivities.size());
                Logger.info("!!!started activities count ", startedActivities.size());
                Logger.info("!!!resumed activities count ", resumedActivities.size());
                Logger.info("!!!paused activities count ", pausedActivities.size());
                Logger.info("!!!stopped activities count ", stoppedActivities.size());
                Logger.info("!!!restarted activities count ", restartedActivities.size());
                Logger.info("!!!destroyed activities count ", destroyedActivities.size());
                if (resumedActivities.iterator().hasNext()) {
                    currentActivity[0] = resumedActivities.iterator().next();
                }
            }
        });

        return currentActivity[0];
    }
```

The output is

```
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!pre on create activities count 0
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!created activities count 0
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!started activities count 0
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!resumed activities count 0
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!paused activities count 0
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!stopped activities count 0
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!restarted activities count 0
01-24 16:02:06.413 9375-9375/io.appium.uiautomator2.e2etest I/appium: !!!destroyed activities count 0
```

Does anyone have any ideas on why the ActivityLifecycleMonitorRegistry isn't finding any activities? Is it possible that I'm running the code from the wrong thread and not the application threads?

**Comment by bootstraponline** [comment]
If that didn't work then you might have to use Selendroid's method for obtaining the activity. They don't use the espresso code.

**Comment by Rovack** [comment]
Does the milestone change imply this is not going to happen within the next month or so after all?

**Comment by dpgraham** [comment]
@Rovack Whenever a milestone finishes, we always push the issue to the next milestone if it hasn't been completed. We're going to look into this issue more and should have an updated timeline soon. It's still high priority.

**Comment by jlipps** [comment]
@sravanmedarapu any more ideas on how we could tackle this?

**Comment by Kureev** [comment]
Looks like this story has been removed from the sprint 8 and wasn't able to sprint 9. Does it mean it's not a high-prio anymore?

**Comment by dpgraham** [comment]
It's still a high priority, but we're blocked right now. We're using Android's UIAutomator2 framework, which doesn't have access to Activities or Views. Myself, and @sravanmedarapu have tried workarounds (the Selendroid workaround above) but haven't had any success. 

Is there anyone on the React Native team with Android experience who could lend their expertise on this?

**Comment by Kureev** [comment]
I think @mkonicek can give you an insightful feedback

**Comment by dpgraham** [comment]
@Kureev Should I reach out to him directly or should I wait for him to contact us?

**Comment by Kureev** [comment]
I sent him a link to the issue. Most likely he's busy now. I believe he'll answer as soon as possible. In the meantime, feel free to ping me if something. I'd like to help solving this issue, but I have no strong Android skills, unfortunately.

**Comment by Rovack** [comment]
So, it's been 2 months... I take it @mkonicek is **really** busy?

**Comment by chaz2k** [comment]
@jlipps @mkonicek Where can I find the source for FindElementsBaseHandler? I'm surprised that it wasn't included along with a pom.xml file.

**Comment by jlipps** [comment]
Update on this: we have implemented a basic [espresso driver for appium](https://github.com/appium/appium-espresso-driver) which will support finding elements by view tag. given that "view tag" is not currently a selenium-compatible locator strategy, we will have to add that strategy to the appium clients before they can be used with this aspect of the espresso driver. when that work is done we'll update here again. all this to say, something's coming!

**Comment by dpgraham** [comment]
If I can add more details to this, the reason we're not able to support finding view tags in our current Android driver, [appium uiautomator2 driver](https://github.com/appium/appium-uiautomator2-server), is because it uses UIAutomator2 as it's underlying technology. UIAutomator2 is strictly a black box automated UI testing framework, and only provides us with an API for the purpose of simulating app usage, and does not provide us with access to the app's view hierarchy. As such, we can't currently support accessing elements by view tags.

**Comment by langme** [comment]
Have you got a other solution ? my goal is to catch TAG with python/appium

**Comment by mkonicek** [comment]
Thanks for the very useful context @dpgraham! Should this issue be closed now?

I posted a short update on the related React Native pull request:
https://github.com/facebook/react-native/pull/9942#issuecomment-362914681

**Comment by bootstraponline** [comment]
uiautomator2 runs in an instrumentation context. espresso can also run there... the find by view id API should work without issue. This would be a hybrid use, where uiautomator2 powers most of the APIs and then Espresso is invoked for this one feature.

**Comment by kakaying** [comment]
Hi, 
Do we have some updates on how to use view tag locating element via espresso driver in Appium? Is this available in Appium 1.7.2 now?

**Comment by mrk-han** [comment]
@jlipps Any luck implementing "view tag" as a selenium-compatible locator strategy in the Appium Client?

Any update on this would be awesome.

**Comment by samarthajoshua** [comment]
Hi, Is there any update on this?

**Comment by dpgraham** [comment]
We're currently, actively working on appium-espresso-driver which will have a view tag selector and we expect it out soon.

Does anyone have an example of a React Native app that uses view tags?

**Comment by bdelville** [comment]
It comes late, but here a simple reac-native debug build containing 3 texts. One with nativeID, one with testID (which is translated to tags), one with accesibilityLabel

https://wetransfer.com/downloads/1cdb4dc661cb288bd9396854774368df20180719021504/192b7d4dba5d521eb06b27a0112e2aef20180719021504/4e2e36

**Comment by dpgraham** [comment]
That's perfect, thanks @bdelville!

**Comment by samarthajoshua** [comment]
@dpgraham Not to be a pain here but how soon is 'soon'? :)

**Comment by dpgraham** [comment]
Here's the compatibility table for Espresso: https://docs.google.com/spreadsheets/d/e/2PACX-1vTNUqxnCl84MYD15UUyQVgO1XjLHOqeWfs-CP--1NwUH9etVCubKSwDVdOa-7rMwDGj6wXGn7Z4qHp9/pubhtml

The Espresso Driver is in beta right now so folks can use it.

As for the View tag, if somebody could give me some assistance on this, as I don't fully understand how view tags work with React.

I'm planning on using the Espresso viewMatcher internally and I'm not sure if I should be using `withTagKey` or `withTagValue` (see https://developer.android.com/reference/android/support/test/espresso/matcher/ViewMatchers#withtagkey)

Would `withTagKey(<tag string>)` be sufficient?

**Comment by mykola-mokhnach** [comment]
Closed as fixed (Espresso driver only though)

**Comment by dpgraham** [comment]
This will be available in the next Appium version and it will be in Appium beta later today.
